src/rdb.c

```diff
@@ -1214,50 +1214,51 @@ ssize_t rdbSaveSingleModuleAux(rio *rdb, int when, moduleType *mt) {
     return io.bytes;
 }
 
-/* Produces a dump of the database in RDB format sending it to the specified
- * Redis I/O channel. On success C_OK is returned, otherwise C_ERR
- * is returned and part of the output, or all the output, can be
- * missing because of I/O errors.
- *
- * When the function returns C_ERR and if 'error' is not NULL, the
- * integer pointed by 'error' is set to the value of errno just after the I/O
- * error. */
-int rdbSaveRio(rio *rdb, int *error, int rdbflags, rdbSaveInfo *rsi) {
-    dictIterator *di = NULL;
-    dictEntry *de;
-    char magic[10];
-    uint64_t cksum;
-    size_t processed = 0;
-    int j;
-    long key_count = 0;
-    long long info_updated_time = 0;
-    char *pname = (rdbflags & RDBFLAGS_AOF_PREAMBLE) ? "AOF rewrite" :  "RDB";
-
-    if (server.rdb_checksum)
-        rdb->update_cksum = rioGenericUpdateChecksum;
-    snprintf(magic,sizeof(magic),"REDIS%04d",RDB_VERSION);
-    if (rdbWriteRaw(rdb,magic,9) == -1) goto werr;
-    if (rdbSaveInfoAuxFields(rdb,rdbflags,rsi) == -1) goto werr;
-    if (rdbSaveModulesAux(rdb, REDISMODULE_AUX_BEFORE_RDB) == -1) goto werr;
-
-    /* save functions */
```

Comment:
I think it'll need dirty the blame in any case because of this `if`. So the idea was to structure the code so that each part of the top level RDB is wrapped in a function which can be easily called or not called based on filter/flag/etc.

Reply:
a bit ugly, but we can add an `if` with a `break` or `continue` inside the db loop.
or cache server.db num on the stack, and add an `if` that overrides it with 0

Reply:
for the record, we discussed that with Yossi and decided to keep the refactory.
but IIRC we said that we'll make a function that saves a single DB, and keep the loop outside.
besides that, please see if you can re-order the functions in a way that will make the diff more "reviewable" (if possible).

Reply:
Did the change hope it's good enough.

Reply:
time will tell (since this is an attempt to avoid other future refactoring).
there are bugs, but i commented on them, so resolving this one.